File: .github/scripts/build-indexes.py
#!/usr/bin/env python3
"""
Rebuild aglets/index.json + plugins/index.json from per-package meta.json
files. Pure scan: globs <ns>/*/meta.json under repo root, projects each
into the catalog row shape, sorts by updated_at desc, writes back.

The output shape mirrors what `aglet publish` / `aglet plugin publish`
previously wrote inline. Triggering this workflow on main push means the
publish PRs themselves no longer have to maintain index.json — they
just drop in their meta.json + tarball, and the index lands here.

Idempotent: if the rebuilt JSON byte-equals what's on disk, the workflow
detects "no diff" and skips the commit.
"""

from __future__ import annotations

import datetime as _dt
import glob
import json
import os
import sys
from typing import Any


REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))


def utc_now_iso() -> str:
    return _dt.datetime.now(_dt.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _load(path: str) -> dict[str, Any]:
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def _copy_fields(meta: dict[str, Any], keys: list[str], dst: dict[str, Any]) -> None:
    """Project the listed meta fields onto dst when present (preserve i18n
    objects, arrays, etc — no flattening)."""
    for k in keys:
        if k in meta:
            dst[k] = meta[k]


def _latest_version_obj(meta: dict[str, Any]) -> dict[str, Any] | None:
    versions = meta.get("versions") or []
    latest_tag = meta.get("latest")
    if not latest_tag:
        return versions[-1] if versions else None
    for v in versions:
        if v.get("version") == latest_tag:
            return v
    # Fallback: latest tag doesn't match any version row.
    return versions[-1] if versions else None


# ─── aglets ──────────────────────────────────────────────────────────────

AGLET_STORE_FIELDS = [
    "name", "description", "author", "homepage", "license", "icon",
    "category", "keywords", "screenshots", "repository",
]


def build_aglet_entry(meta_path: str) -> dict[str, Any]:
    meta = _load(meta_path)
    entry: dict[str, Any] = {"id": meta["id"]}
    _copy_fields(meta, AGLET_STORE_FIELDS, entry)
    entry["latest"] = meta.get("latest", "")
    latest_v = _latest_version_obj(meta)
    if latest_v and "published_at" in latest_v:
        entry["updated_at"] = latest_v["published_at"]
    return entry


def build_aglets_index() -> dict[str, Any]:
    metas = sorted(glob.glob(os.path.join(REPO_ROOT, "aglets", "*", "meta.json")))
    entries = [build_aglet_entry(p) for p in metas]
    entries.sort(key=lambda e: e.get("updated_at", ""), reverse=True)
    return {"generated_at": utc_now_iso(), "aglets": entries}


# ─── plugins ─────────────────────────────────────────────────────────────

PLUGIN_STORE_FIELDS = [
    "name", "description", "category", "keywords", "author",
    "homepage", "repository", "license",
]


def build_plugin_entry(meta_path: str) -> dict[str, Any]:
    meta = _load(meta_path)
    entry: dict[str, Any] = {"id": meta["id"]}
    _copy_fields(meta, PLUGIN_STORE_FIELDS, entry)
    # Plugin-specific runtime metadata (mirrors aglet plugin publish output):
    if "namespace" in meta:
        entry["namespace"] = meta["namespace"]
    if "backend_kind" in meta:
        entry["backend_kind"] = meta["backend_kind"]
    if "wasm_features" in meta:
        entry["wasm_features"] = meta["wasm_features"]
    actions = meta.get("actions") or []
    entry["actions_count"] = len(actions)
    entry["latest"] = meta.get("latest", "")
    latest_v = _latest_version_obj(meta)
    if latest_v:
        if "wasm_size" in latest_v:
            entry["wasm_size"] = latest_v["wasm_size"]
        if "published_at" in latest_v:
            entry["updated_at"] = latest_v["published_at"]
    return entry


def build_plugins_index() -> dict[str, Any]:
    metas = sorted(glob.glob(os.path.join(REPO_ROOT, "plugins", "*", "meta.json")))
    entries = [build_plugin_entry(p) for p in metas]
    entries.sort(key=lambda e: e.get("updated_at", ""), reverse=True)
    return {"generated_at": utc_now_iso(), "plugins": entries}


# ─── write back ──────────────────────────────────────────────────────────

def write_json(path: str, obj: dict[str, Any]) -> None:
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(obj, f, indent=2, ensure_ascii=False)
        f.write("\n")


def main() -> int:
    aglets_idx = build_aglets_index()
    plugins_idx = build_plugins_index()
    write_json(os.path.join(REPO_ROOT, "aglets", "index.json"), aglets_idx)
    write_json(os.path.join(REPO_ROOT, "plugins", "index.json"), plugins_idx)
    print(f"aglets/index.json:  {len(aglets_idx['aglets'])} entries")
    print(f"plugins/index.json: {len(plugins_idx['plugins'])} entries")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .github/scripts/build-indexes.ts
#!/usr/bin/env bun
/**
 * Rebuild aglets/index.json + plugins/index.json from per-package meta.json
 * files. Pure scan: globs <ns>/*​/meta.json under repo root, projects each
 * into the catalog row shape, sorts by updated_at desc, writes back.
 *
 * The output shape mirrors what `aglet publish` / `aglet plugin publish`
 * write into each package's meta.json. Triggering this on main push means
 * publish PRs don't maintain index.json — they drop in meta.json + tarball,
 * and the index lands here.
 *
 * Idempotent: rebuilt JSON byte-matches the prior output (modulo generated_at)
 * → workflow's git-diff check skips the commit when nothing changed.
 *
 * TS (run via bun) —— 仓库禁用 Python。JSON 输出对齐旧 build-indexes.py
 * （indent=2、ensure_ascii=False ≙ JSON.stringify(…,2) + UTF-8、尾随换行）。
 */

import {
  readdirSync,
  readFileSync,
  writeFileSync,
  mkdirSync,
  existsSync,
} from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const REPO_ROOT = join(dirname(fileURLToPath(import.meta.url)), "..", "..");

type Json = Record<string, any>;

function utcNowIso(): string {
  // 对齐 Python strftime("%Y-%m-%dT%H:%M:%SZ") —— 去掉毫秒。
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function load(path: string): Json {
  return JSON.parse(readFileSync(path, "utf-8"));
}

/** Project listed meta fields onto dst when present (preserve i18n objects, arrays). */
function copyFields(meta: Json, keys: string[], dst: Json): void {
  for (const k of keys) if (k in meta) dst[k] = meta[k];
}

function latestVersionObj(meta: Json): Json | null {
  const versions: Json[] = meta.versions ?? [];
  const latestTag = meta.latest;
  if (!latestTag) return versions.length ? versions[versions.length - 1] : null;
  for (const v of versions) if (v.version === latestTag) return v;
  // Fallback: latest tag doesn't match any version row.
  return versions.length ? versions[versions.length - 1] : null;
}

/** Sorted list of <ns>/<id>/meta.json paths under the repo root. */
function metaPaths(ns: string): string[] {
  const base = join(REPO_ROOT, ns);
  if (!existsSync(base)) return [];
  const out: string[] = [];
  for (const dirent of readdirSync(base, { withFileTypes: true })) {
    if (!dirent.isDirectory()) continue;
    const p = join(base, dirent.name, "meta.json");
    if (existsSync(p)) out.push(p);
  }
  return out.sort();
}

// ─── aglets ──────────────────────────────────────────────────────────────

const AGLET_STORE_FIELDS = [
  "name", "description", "author", "homepage", "license", "icon",
  "category", "keywords", "screenshots", "repository",
  // platforms: aglet 支持的原生平台集（["apple","android","windows"]），由
  // `aglet publish` 经 spec.pickRenderer 算入 meta.json。Store(macOS) /
  // catalog(Android) 据此「只列支持本平台的 aglet」（不向后兼容：缺失即不列出，
  // 每个 aglet 都应以含本字段的新 CLI 重新发布）。
  "platforms",
];

function buildAgletEntry(metaPath: string): Json {
  const meta = load(metaPath);
  const entry: Json = { id: meta.id };
  copyFields(meta, AGLET_STORE_FIELDS, entry);
  entry.latest = meta.latest ?? "";
  const latestV = latestVersionObj(meta);
  if (latestV && "published_at" in latestV) entry.updated_at = latestV.published_at;
  return entry;
}

function buildAgletsIndex(): Json {
  const entries = metaPaths("aglets").map(buildAgletEntry);
  entries.sort((a, b) =>
    String(b.updated_at ?? "").localeCompare(String(a.updated_at ?? "")));
  return { generated_at: utcNowIso(), aglets: entries };
}

// ─── plugins ─────────────────────────────────────────────────────────────

const PLUGIN_STORE_FIELDS = [
  "name", "description", "category", "keywords", "author",
  "homepage", "repository", "license",
];

function buildPluginEntry(metaPath: string): Json {
  const meta = load(metaPath);
  const entry: Json = { id: meta.id };
  copyFields(meta, PLUGIN_STORE_FIELDS, entry);
  // Plugin-specific runtime metadata (mirrors aglet plugin publish output):
  if ("namespace" in meta) entry.namespace = meta.namespace;
  if ("backend_kind" in meta) entry.backend_kind = meta.backend_kind;
  if ("wasm_features" in meta) entry.wasm_features = meta.wasm_features;
  const actions = meta.actions ?? [];
  entry.actions_count = actions.length;
  entry.latest = meta.latest ?? "";
  const latestV = latestVersionObj(meta);
  if (latestV) {
    if ("wasm_size" in latestV) entry.wasm_size = latestV.wasm_size;
    if ("published_at" in latestV) entry.updated_at = latestV.published_at;
  }
  return entry;
}

function buildPluginsIndex(): Json {
  const entries = metaPaths("plugins").map(buildPluginEntry);
  entries.sort((a, b) =>
    String(b.updated_at ?? "").localeCompare(String(a.updated_at ?? "")));
  return { generated_at: utcNowIso(), plugins: entries };
}

// ─── write back ──────────────────────────────────────────────────────────

function writeJson(path: string, obj: Json): void {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, JSON.stringify(obj, null, 2) + "\n");
}

function main(): void {
  const agletsIdx = buildAgletsIndex();
  const pluginsIdx = buildPluginsIndex();
  writeJson(join(REPO_ROOT, "aglets", "index.json"), agletsIdx);
  writeJson(join(REPO_ROOT, "plugins", "index.json"), pluginsIdx);
  console.log(`aglets/index.json:  ${agletsIdx.aglets.length} entries`);
  console.log(`plugins/index.json: ${pluginsIdx.plugins.length} entries`);
}

main();
